easypy/lockstep.py

```diff
@@ -0,0 +1,104 @@
+
+        try:
+            self._current_step, value = self._next_step_and_value()
+        except StopIteration:
+            raise LockstepSyncMismatch(lockstep_name=self._name,
+                                       expected_step=step,
+                                       actual_step='finished')
+        if self._current_step != step:
+            raise LockstepSyncMismatch(lockstep_name=self._name,
+                                       expected_step=step,
+                                       actual_step=self._current_step)
+        return value
+
+    def step_until(self, step):
+        """Progress until we get to the specified step, confirming that it exist"""
+
+        while self._current_step != step:
+            try:
+                self._current_step, value = self._next_step_and_value()
+            except StopIteration:
+                raise LockstepSyncMismatch(lockstep_name=self._name,
+                                           expected_step=step,
+                                           actual_step='finished')
+        return value
+
+    def step_all(self):
+        """Progress through all the remaining steps"""
+
+        for self._current_step in self._generator:
+            pass
+
+    def _next_step_and_value(self):
+        yield_result = next(self._generator)
+        if isinstance(yield_result, tuple) and len(yield_result) == 2:
+            return yield_result
+        else:
+            return yield_result, None
+
+    def __str__(self):
+        return '%s<%s>' % (self._name, self._current_step)
+
+
+@contextmanager
+def __fake_cm():
+    yield
+
+
+class _LockstepGeneratorContextManager(type(__fake_cm())):
+    def __iter__(self):
+        with self as invocation:
+            yield from invocation._generator
+
+
+def lockstep(generator_func):
+    """
+    Synchronize a coroutine that runs a process step-by-step.
+
+    Decorate a generator that yields step names to create a context manager.
+    The context object has a `.step_next`/`.step_until` methods that must be
+    called, in order, with all expected step names, to make the generator
+    progress to each step.
+    """
+
+    @contextmanager
+    def cm(*args, **kwargs):
+        invocation = _LockstepInvocation(generator_func.__name__, generator_func(*args, **kwargs))
+
+        yield invocation
+
+        try:
+            step_not_taken, _ = invocation._next_step_and_value()
+        except StopIteration:
+            # all is well - all steps were exhausted
+            invocation._current_step = 'finished'
+        else:
+            raise LockstepSyncMismatch(lockstep_name=invocation._name,
+                                       expected_step='finished',
+                                       actual_step=step_not_taken)
+
+    @wraps(generator_func)
+    def wrapper(*args, **kwargs):
+        result = cm(*args, **kwargs)
+        result.__class__ = _LockstepGeneratorContextManager
```

Comment:
@koreno - sorry for this ugly hack, but...

Somewhere between Python 3.4.2 and 3.4.5 they changed `_GeneratorContextManager.__init__` from
```python
class _GeneratorContextManager(ContextDecorator):
    """Helper for @contextmanager decorator."""

    def __init__(self, func, *args, **kwds):
```

to
```python
class _GeneratorContextManager(ContextDecorator):
    """Helper for @contextmanager decorator."""

    def __init__(self, func, args, kwds):
```

This means we can't use the same code to initiate both versions. Well, it was a private interface to begin with...

So instead, I'll use `contextmanager` to create the object, and switch it's class to something that inherits what `contextmanager` usually returns.